File: apps/ltx-studio/shared/runtimeTrust.ts
import { z } from "zod";

const sha256Schema = z.string().regex(/^[0-9a-f]{64}$/);

export const authorityIsolationSchema = z.discriminatedUnion("status", [
  z.object({
    schemaVersion: z.literal("ltx-studio-authority-isolation.v1"),
    status: z.literal("hold"),
    mechanism: z.literal("same-local-uid"),
    attestationSha256: z.null(),
    reasonCode: z.literal("same-uid-authority-not-authentic"),
  }).strict(),
  z.discriminatedUnion("mechanism", [
    z.object({
      schemaVersion: z.literal("ltx-studio-authority-isolation.v1"),
      status: z.literal("attested"),
      mechanism: z.literal("separate-studio-identity-proc-fd-isolation"),
      hostTcbAttestationSha256: sha256Schema,
      brokerAttestationSha256: z.null(),
      reasonCode: z.null(),
    }).strict(),
    z.object({
      schemaVersion: z.literal("ltx-studio-authority-isolation.v1"),
      status: z.literal("attested"),
      mechanism: z.literal("external-signer-sealed-fd-broker"),
      hostTcbAttestationSha256: sha256Schema,
      brokerAttestationSha256: sha256Schema,
      reasonCode: z.null(),
    }).strict(),
  ]),
]);

export type AuthorityIsolation = z.infer<typeof authorityIsolationSchema>;

/**
 * The parts of a production identity that live outside the immutable release
 * tree.  Keeping this as one strict object prevents an authorization from
 * accidentally binding only the in-release runtime seal while omitting the
 * host, service or build trust anchors.
 */
export const runtimeTrustBindingSchema = z.object({
  schemaVersion: z.literal("ltx-studio-runtime-trust-binding.v2"),
  hostTcbAttestationSha256: sha256Schema,
  hostTcbContractSha256: sha256Schema,
  servicePolicySha256: sha256Schema,
  buildTcbSha256: sha256Schema,
  authorityIsolation: authorityIsolationSchema,
  trustPolicyDigests: z.object({
    release: sha256Schema,
    activationWriter: sha256Schema,
    qualificationAuthorizer: sha256Schema,
    runtimeRights: sha256Schema,
    bootstrapAuthority: sha256Schema,
  }).strict(),
}).strict().superRefine((runtimeTrust, context) => {
  if (runtimeTrust.authorityIsolation.status === "attested"
    && runtimeTrust.authorityIsolation.hostTcbAttestationSha256
      !== runtimeTrust.hostTcbAttestationSha256) {
    context.addIssue({
      code: "custom",
      path: ["authorityIsolation", "hostTcbAttestationSha256"],
      message: "authority isolation must be attested by the exact pinned Host-TCB record",
    });
  }
});

export type RuntimeTrustBinding = z.infer<typeof runtimeTrustBindingSchema>;

export function assertRuntimeTrustAuthorizesRelease(
  value: unknown,
  context = "Release",
): RuntimeTrustBinding {
  const runtimeTrust = runtimeTrustBindingSchema.parse(value);
  if (runtimeTrust.authorityIsolation.status !== "attested") {
    throw new Error(
      `${context} blocked: same-local-UID execution/publication authority is not authentic; `
      + "a separately attested Studio identity with proc/fd isolation or an external signer/sealed-FD broker is required",
    );
  }
  return runtimeTrust;
}

export const runtimeIdentityBindingSchema = z.object({
  releaseDigest: sha256Schema,
  surfaceDigest: sha256Schema,
  runtimeInstallSealSha256: sha256Schema,
  runtimeTreeSha256: sha256Schema,
  runtimePolicySha256: sha256Schema,
  nodeExecutableSha256: sha256Schema,
  runtimeTrust: runtimeTrustBindingSchema,
}).strict();

export type RuntimeIdentityBinding = z.infer<typeof runtimeIdentityBindingSchema>;

File: packages/ltx-trainer/src/ltx_trainer/av_eval/runtime_trust.py
"""Strict validation for Studio RuntimeTrust bindings used by AV evaluation."""

from __future__ import annotations

import copy
from typing import Any

RUNTIME_TRUST_SCHEMA = "ltx-studio-runtime-trust-binding.v2"
AUTHORITY_ISOLATION_SCHEMA = "ltx-studio-authority-isolation.v1"
ATTESTED_AUTHORITY_MECHANISMS = {
    "external-signer-sealed-fd-broker",
    "separate-studio-identity-proc-fd-isolation",
}


class RuntimeTrustBindingError(ValueError):
    """Raised when an external Studio RuntimeTrust binding is absent or unsafe."""


def _exact_keys(value: dict[str, Any], expected: set[str], context: str) -> None:
    missing = sorted(expected - value.keys())
    unknown = sorted(value.keys() - expected)
    if missing or unknown:
        raise RuntimeTrustBindingError(f"{context}: missing={missing}, unknown={unknown}")


def _sha256(value: object, context: str) -> str:
    if (
        not isinstance(value, str)
        or len(value) != 64
        or any(character not in "0123456789abcdef" for character in value)
    ):
        raise RuntimeTrustBindingError(f"{context} must be a lowercase SHA-256")
    return value


def validate_runtime_trust_binding(raw: object) -> dict[str, Any]:
    """Return a defensive copy of an exact, release-capable RuntimeTrust v2 binding.

    This validates the document contract shared with ``runtimeTrustBindingSchema``
    in LTX Studio.  It deliberately does not synthesize any digest.  The caller
    must additionally bind the returned value to independently signed evidence.
    """

    if not isinstance(raw, dict):
        raise RuntimeTrustBindingError("qualification HOLD: an externally verified RuntimeTrust binding is required")
    _exact_keys(
        raw,
        {
            "schemaVersion",
            "hostTcbAttestationSha256",
            "hostTcbContractSha256",
            "servicePolicySha256",
            "buildTcbSha256",
            "authorityIsolation",
            "trustPolicyDigests",
        },
        "RuntimeTrust binding",
    )
    if raw["schemaVersion"] != RUNTIME_TRUST_SCHEMA:
        raise RuntimeTrustBindingError("qualification HOLD: unsupported RuntimeTrust binding schema")
    for field in (
        "hostTcbAttestationSha256",
        "hostTcbContractSha256",
        "servicePolicySha256",
        "buildTcbSha256",
    ):
        _sha256(raw[field], f"RuntimeTrust.{field}")

    policies = raw["trustPolicyDigests"]
    if not isinstance(policies, dict):
        raise RuntimeTrustBindingError("RuntimeTrust.trustPolicyDigests must be an object")
    _exact_keys(
        policies,
        {
            "release",
            "activationWriter",
            "qualificationAuthorizer",
            "runtimeRights",
            "bootstrapAuthority",
        },
        "RuntimeTrust.trustPolicyDigests",
    )
    for role, digest in policies.items():
        _sha256(digest, f"RuntimeTrust.trustPolicyDigests.{role}")

    isolation = raw["authorityIsolation"]
    if not isinstance(isolation, dict):
        raise RuntimeTrustBindingError("RuntimeTrust.authorityIsolation must be an object")
    if isolation.get("status") == "hold":
        _exact_keys(
            isolation,
            {"schemaVersion", "status", "mechanism", "attestationSha256", "reasonCode"},
            "RuntimeTrust.authorityIsolation",
        )
        if (
            isolation["schemaVersion"] != AUTHORITY_ISOLATION_SCHEMA
            or isolation["mechanism"] != "same-local-uid"
            or isolation["attestationSha256"] is not None
            or isolation["reasonCode"] != "same-uid-authority-not-authentic"
        ):
            raise RuntimeTrustBindingError("RuntimeTrust authority HOLD contract is invalid")
        raise RuntimeTrustBindingError(
            "qualification HOLD: RuntimeTrust authority isolation is not externally attested"
        )
    _exact_keys(
        isolation,
        {
            "schemaVersion",
            "status",
            "mechanism",
            "hostTcbAttestationSha256",
            "brokerAttestationSha256",
            "reasonCode",
        },
        "RuntimeTrust.authorityIsolation",
    )
    if (
        isolation["schemaVersion"] != AUTHORITY_ISOLATION_SCHEMA
        or isolation["status"] != "attested"
        or isolation["mechanism"] not in ATTESTED_AUTHORITY_MECHANISMS
        or isolation["reasonCode"] is not None
    ):
        raise RuntimeTrustBindingError(
            "qualification HOLD: RuntimeTrust authority isolation is not a supported attestation"
        )
    _sha256(
        isolation["hostTcbAttestationSha256"],
        "RuntimeTrust.authorityIsolation.hostTcbAttestationSha256",
    )
    if isolation["hostTcbAttestationSha256"] != raw["hostTcbAttestationSha256"]:
        raise RuntimeTrustBindingError("RuntimeTrust authority isolation does not bind the exact Host-TCB attestation")
    if isolation["mechanism"] == "external-signer-sealed-fd-broker":
        _sha256(
            isolation["brokerAttestationSha256"],
            "RuntimeTrust.authorityIsolation.brokerAttestationSha256",
        )
    elif isolation["brokerAttestationSha256"] is not None:
        raise RuntimeTrustBindingError("separate Studio identity RuntimeTrust must not claim a broker attestation")
    return copy.deepcopy(raw)
